nntrainer/tensor/cl_operations/attention_kernel_interface.cpp

```diff
@@ -0,0 +1,136 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2024 Yash Singh <yash.singh@samsung.com>
+ *
+ * @file	attention_kernel_interface.cpp
+ * @date	28 August 2024
+ * @brief	Interface for attention OpenCL kernels
+ * @see		https://github.com/nnstreamer/nntrainer
+ * @author	Yash Singh <yash.singh@samsung.com>
+ * @bug		No known bugs except for NYI items
+ *
+ */
+
+#include <attention_kernel_interface.h>
+#include <attention_kernels.h>
+
+namespace nntrainer {
+/**
+ * @brief      compute frequency for rotary embedding
+ * @param[in]  dim hidden dim size
+ * @param[in]  seq_len sequency length
+ * @param[out] freqs_cos cosine of the frequencies
+ * @param[out] freqs_sin sine of the frequencies
+ * @param[out] freqs base frequencies array to be used in the future computation
+ * @param[in]  theta rotary angle
+ */
+void precompute_freqs(int dim, unsigned int seq_len,
+                      std::vector<std::vector<float>> &freqs_cos,
+                      std::vector<std::vector<float>> &freqs_sin,
+                      std::vector<float> &freqs, float theta = 10000.0) {
+  unsigned int half_ = dim / 2;
+  for (unsigned int i = 0; i < half_; ++i) {
+    freqs.push_back(1.0 / (std::pow(theta, (2 * i) / static_cast<float>(dim))));
+  }
+
+  auto cos = std::vector<std::vector<float>>();
+  cos.assign(seq_len, std::vector<float>(dim, 0));
+
+  auto sin = std::vector<std::vector<float>>();
+  sin.assign(seq_len, std::vector<float>(dim, 0));
+
+  for (unsigned int i = 0; i < seq_len; ++i) {
+    for (unsigned int j = 0; j < half_; ++j) {
+      float angle = i * freqs[j];
+      cos[i][j] = std::cos(angle);
+      cos[i][j + half_] = std::cos(angle); // repeated 2 times
+
+      sin[i][j] = std::sin(angle);
+      sin[i][j + half_] = std::sin(angle); // repeated 2 times
+    }
+  }
+  freqs_cos = cos;
+  freqs_sin = sin;
+}
+
+/**
+ * @brief     apply rotary embedding
+ * @param[in] in input tensor
+ * @param[in] dim hidden dim size
+ * @param[in] from sequence order
+ * @param[in] max_timestep maximum timestep
+ */
+void apply_rotary_emb_cl(Tensor &in, unsigned int dim, unsigned int from,
+                         unsigned int max_timestep, RunLayerContext &context) {
+  nntrainer::Tensor out(in.getDim());
+  float value = 0;
+  float transformed_value = 0.0;
```

Comment:
it is not a big deal, but it would be better to initialize it with 0.0f. ([reference](https://stackoverflow.com/questions/5199338/what-is-the-significance-of-0-0f-when-initializing-in-c))
```suggestion
  float value = 0.0f;
  float transformed_value = 0.0f;
```